### PR TITLE
first commit for public/private key signing JDK

### PR DESCRIPTION
**Pull request by bblfish**
This is a first attempt to get something working in the public/private key space.
The 1 test for the JDK works, but does not pass. I am not sure if that is the IETF "Signing HTTP Messages" spec that gives the wrong input or the code that is wrong. 

There is still a lot to do. But this is really to start a bit of a discussion.

### CODE REVIEW

**Comment by bblfish** [comment]
Going through the code, I wrote up a number of questions, as they occurred to me.  I also placed some comments in the code.  I am very new to this so the answers may be quite simple or obvious, but then that could be useful for documentation.

1.  the SecretKeySpec is a pair of `ByteVector` and `Algorithm`.  For java those are turned into a `SecretKeySpec`. I wonder if the transformation in and out of the java object is costly (may not be). If it were would an opaque type be better? Probably premature optimization. Also it looks like one could create illegal objects this way, by just choosing an arbitrary array of bytes and an encoding.  Actually this efficiency question becomes clearer with `X509EncodedKeySpec` which have to parse the byte string into an object. Since those are immutable, it is wasteful to have to parse this on every signing.
2. Should the Algorithms allowed for `SecretKeySpec` be a subset of the Algorithms? Perhaps `SecretAlgorithms`?
3. there is no test for `HMac.importKey`
4. `Crypto[F]`
   * There are no tests
   * should that perhaps be called `HmacCrypto` ?
5. Given how new the project is would it not be best to write it in Scala 3 with x-compilation to 2.13  ?

**Comment by bblfish** [comment]
Ah yes to try the test you have to be in project `coreJVM`

**Comment by armanbilge** [comment]
Thanks for working on this! Let me try and answer some of your questions.

1. To start, I'd like to stick with the `ByteVector`/`Algorithm` encoding, because this is guaranteed to be cross-platform. I agree it will have some performance costs due to the conversion, so we can explore solutions for that problem later.
2. Yes, that sounds good 👍 
3. Good catch. Also you make a good point about illegal keys. Maybe we can try factoring out the `importKey`, it's not specific to `Hmac` I believe.
4. What tests do you think we should add for `Crypto`? It is not intended to be focused on `Hmac`, it is meant to be a catch-all capability trait for cryptography (i.e., all the APIs should be available through it).
5. Is cross-compiling for 2.12 hampering the design in some way? I'm open to dropping it.

File: build.sbt [hunk 1]
```diff
@@ -106,9 +107,12 @@ lazy val core = crossProject(JSPlatform, JVMPlatform)
       "org.typelevel" %%% "cats-laws" % catsVersion % Test,
       "org.typelevel" %%% "cats-effect" % catsEffectVersion % Test,
       "org.typelevel" %%% "discipline-munit" % disciplineMUnitVersion % Test,
-      "org.typelevel" %%% "munit-cats-effect-3" % munitCEVersion % Test
+      "org.typelevel" %%% "munit-cats-effect-3" % munitCEVersion % Test,
     )
   )
+  .jvmSettings(
+    libraryDependencies +="org.bouncycastle" % "bcpkix-jdk15to18" % bouncyVersion % Test
```
Comment (armanbilge) [inline build.sbt:114]: Is this dependency required to implement these algorithms? Is it possible to implement these algorithms in Node.js and browsers?

File: core/jvm/src/main/scala/bobcats/KeyPlatform.scala [hunk 2]
```diff
@@ -16,19 +16,17 @@
 
 package bobcats
 
-import java.security
 import javax.crypto
 
-private[bobcats] trait KeyPlatform {
-  def toJava: security.Key
-}
```
Comment (armanbilge) [inline core/jvm/src/main/scala/bobcats/KeyPlatform.scala:24]: Why did this change? A `KeySpec` is different from a `Key`, and I think it makes sense to have `Key` here. That way we don't need to decide which `KeySpec` to use.

File: build.sbt (hunk 1 above)
Comment (bblfish) [reply build.sbt:114]: I don't think so. 
I am using examples from the [Signing HTTP Messages](https://httpwg.org/http-extensions/draft-ietf-httpbis-message-signatures.html) spec referenced in the test suite, and last time I tried I felt like I had to reach out to bouncy-castle. As I am building this for the [an implementation of that spec](https://github.com/bblfish/httpSig), I want to make sure it works. If we can find a way to drop that, I'd love to know.

File: core/jvm/src/main/scala/bobcats/KeyPlatform.scala (hunk 2 above)
Comment (bblfish) [reply core/jvm/src/main/scala/bobcats/KeyPlatform.scala:24]: I was having difficulty compiling things, and I thought the reason was a clash here. I'll look at it more carefully again, and see what the problem was.

File: build.sbt (hunk 1 above)
Comment (armanbilge) [reply build.sbt:114]: I'm sorry, I wasn't paying close attention when I first looked. I see now that it is a test-only dependency, which is less problematic. But, any tests we write using it we won't be able to re-use on JavaScript which is unfortunate.

**Comment by bblfish** [comment]
> 4. What tests do you think we should add for Crypto? It is not intended to be focused on Hmac, it is meant to be a catch-all capability trait for cryptography (i.e., all the APIs should be available through it).

I don't understand how it is meant to work. It is defined as

```scala
ealed trait Crypto[F[_]] {
  def hash: Hash[F]
  def hmac: Hmac[F]
}
```
So I am not sure how it is not tied to `Hmac`. How would it work with public key cryptography?

>5.  Is cross-compiling for 2.12 hampering the design in some way? I'm open to dropping it.

Well optimisations that would make use of opaque objects available in Scala 3, can't be implemented If you stick to `2.12`. But if you don't need `2.12` then I think you could move the code to use Scala 3 and compile it down to `2.12`.

**Comment by armanbilge** [comment]
> I don't understand how it is meant to work.

We can keep adding new capabilities to it :) Like this:
```scala
sealed trait Crypto[F[_]] {
  def hash: Hash[F]
  def hmac: Hmac[F]
  def signer: Signer[F]
  def verifier: Verifier[F]
}
```

> Well optimisations that would make use of opaque objects available in Scala 3

Scala 2.13 doesn't have opaque types, so it's not clear to me how this would work.

File: core/jvm/src/main/scala/bobcats/KeyPlatform.scala (hunk 2 above)
Comment (bblfish) [reply core/jvm/src/main/scala/bobcats/KeyPlatform.scala]: ok I fixed this. The `toJava` is back.

**Comment by bblfish** [comment]
I added the tests (finally). 

What those tests show I think, is that one has to be able to fine-tune some algorithms as shown in the  `AlgorithmPlatform.scala` file.  The way I did it there can't be quite right though.

I can't get the signature from the "Signing HTTP Messages" spec in the tests to give me the expected result. I'll try adding more tests and see if I have any luck with the other examples.

**Comment by armanbilge** [comment]
> is that one has to be able to fine-tune some algorithms as shown

Can you clarify what you mean by this?

**Comment by bblfish** [comment]
> > is that one has to be able to fine-tune some algorithms as shown
> 
> Can you clarify what you mean by this?

Ah yes of course, I should have pointed to the code, where I left some comments.
https://github.com/bblfish/bobcats/blob/ca4beabaaa45e9b8d5ce311b34ef02a161a184b6/core/jvm/src/main/scala/bobcats/AlgorithmPlatform.scala#L25

That is because of the definition in the [Signing HTTP Messages spec]( https://httpwg.org/http-extensions/draft-ietf-httpbis-message-signatures.html#section-3.3.1).

**Comment by armanbilge** [comment]
Thanks for clarifying! It seems to me like this information should be stored with the key itself, or directly encoded in the types themselves. I'm looking at the docs for Node.js crypto implementation, where their `KeyObject` has an `asymmetricKeyType`: https://nodejs.org/api/crypto.html#keyobjectasymmetrickeytype

**Comment by bblfish** [comment]
I am finding it difficult to get the tests from the spec to pass. It could be an error in the spec, but it could also be an error in my code.
I will implement the signature verification to see if I can at least verify the signatures I am generating.

**Comment by bblfish** [comment]
I noticed that in your tests you have code to test inside the `map`. 
```scala
   Hmac[F].digest(SecretKeySpec(key, algorithm), data).map { obtained =>
        assertEquals(
          obtained,
          ByteVector.fromHex(expect).get
        )
      }
```     
But if the Functor contains a Sum and maps only occurs on one side of the disjunct, then a failure won't be picked up. So I think Sync works that way.  Any ideas on how to capture the error s being a mistake too?

**Comment by bblfish** [comment]
I added a signature verifier and those do verify the signatures generated. 
So that would tend to indicate the signatures in the spec are wrong, but not necessarily. Perhaps there is a subtle difference in what is being signed.

**Comment by bblfish** [comment]
So I think what needs to be done next is to have a JS version of the code. 
If we find the same problems there then that would give a very good reason to open an issue on the HTTPBis WG GitHub repo. 

Just wondering: is that something you want to do @armanbilge? Perhaps you have all the JS crypto swapped into brain-memory so that it could be done really quickly.

**Comment by armanbilge** [comment]
@bblfish sounds good! I won't have time to look into JS implementations, but they are generally straightforward. Here are the relevant docs:

Node.js: https://nodejs.org/api/crypto.html
Browsers: https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API

You can experiment with them in vanilla JavaScript before to make sure they work as expected before creating a Scala.js facade.

**Comment by bblfish** [comment]
Nearly all tests pass now. 
I have a problem I think with the `§3.1_Signature example` that is consistent across JS and Java, so that could be a problem in the spec.
There is a problem that only appears in JS, so that could be configuration problems.

**Comment by armanbilge** [comment]
Very nice work! CI is failing b/c of formatting (running `scalafmtAll` should fix it) so unfortunately I can't see these specific problems you are experiencing in the logs.

**Comment by bblfish** [comment]
I made some more changes and scalafmt the code.

**Comment by bblfish** [comment]
Ah it looks like I did not scalafmt everything.... Mhh.

**Comment by bblfish** [comment]
I got myself tangled up in git rebase tangle, so I opened instead a new PR https://github.com/typelevel/bobcats/pull/48

**Pull request closed without merging**